src/common/encoding.cpp

```diff
@@ -261,6 +261,82 @@ std::string from_utf8(const std::string& input, eFileEncoding toEncoding)
     return ret;
 }
 
+#else
+
+#include <cerrno>
+#include <memory>
+#include <type_traits>
+#include <vector>
+
+#include <iconv.h>
+
+static const char* get_iconv_encoding_name(eFileEncoding encoding)
+{
+    switch (encoding) {
+    case eFileEncoding::LATIN1:
+        return "ISO-8859-1";
+    case eFileEncoding::SHIFT_JIS:
+        return "CP932";
+    case eFileEncoding::EUC_KR:
+        return "CP949";
+    case eFileEncoding::UTF8:
+        return "UTF-8";
+    }
+    panic("Error", "Incorrect eFileEncoding");
+    abort();
+    return "";
+}
+
+static std::string convert(const std::string& input, eFileEncoding from, eFileEncoding to)
+{
+    const auto* source_encoding_name = get_iconv_encoding_name(from);
+    const auto* target_encoding_name = get_iconv_encoding_name(to);
+
+    auto icd_deleter = [](iconv_t icd) {
+        iconv_close(icd);
+    };
+
+    auto icd = std::unique_ptr<std::remove_pointer<iconv_t>::type, decltype(icd_deleter)>(iconv_open(target_encoding_name, source_encoding_name), icd_deleter);
+    if (reinterpret_cast<size_t>(icd.get()) == static_cast<size_t>(-1)) {
+        panic("Error", "Couldn't open iconv descriptor");
+    }
+
+    static constexpr size_t BUF_SIZE = 1024l * 32l;
+    // I wanted to avoid manually allocating here so that we don't have
+    // to clean up manually in all return paths.
+    char out_buf[BUF_SIZE] = { 0 };
+
+    // BRUH-cast.
+    char* buf_ptr = const_cast<char*>(input.c_str());
+    std::size_t buf_len = input.length();
+    char* out_ptr = static_cast<char*>(out_buf);
+    std::size_t out_len = sizeof(out_buf);
+
+    std::size_t ret_len = iconv(icd.get(), &buf_ptr, &buf_len, &out_ptr, &out_len);
+    if (ret_len == static_cast<size_t>(-1)) {
+        int err_bak = errno;
+        LOG_ERROR << "iconv error: " << err_bak << " (" << strerror(err_bak) << ")";
```

Comment:
strerror is not thread safe, we need to either have a safe wrapper for it that uses a mutex, or use something like pushToMainThreadAndWait() here.

Reply:
> pushToMainThreadAndWait

This won't make sense here, as the problem is that the next function call may modify the underlying buffer, and just delaying the call is not a problem.
Since this code is for non-Win32 only, I wanted to use POSIX's strerror_r, but it works differently on GNU systems. I still think it's the best way to go, so I'll work around this with function overloading.

Reply:
A bit ugly but works: 6f2a317921c63df2bb22486e59824016f5136341